File: src/main/java/net/flyingfishflash/loremlist/domain/exceptions/EntityNotFoundException.java
package net.flyingfishflash.loremlist.domain.exceptions;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.util.Map;
import java.util.Set;
import java.util.UUID;
import java.util.stream.Collectors;
import net.flyingfishflash.loremlist.domain.LrmComponentType;
import org.springframework.http.HttpStatus;

public abstract class EntityNotFoundException extends DomainException {

  public static final HttpStatus HTTP_STATUS = HttpStatus.NOT_FOUND;

  private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();

  protected EntityNotFoundException(
      Set<UUID> idCollection, String message, LrmComponentType lrmComponentType) {
    super(
        null,
        HTTP_STATUS,
        message != null ? message : defaultMessage(idCollection, lrmComponentType),
        null,
        (lrmComponentType != null ? lrmComponentType.name() : "Entity") + "NotFoundException",
        null,
        Map.of("notFound", toJsonElement(idCollection)));
  }

  public static String defaultMessage(Set<UUID> idCollection, LrmComponentType lrmComponentType) {
    String entityName = lrmComponentType != null ? lrmComponentType.name() : "Entity";
    if (idCollection.size() > 1) {
      return entityName + "s (" + idCollection.size() + ") could not be found.";
    }
    return entityName + " could not be found.";
  }

  private static JsonNode toJsonElement(Set<UUID> idCollection) {
    return OBJECT_MAPPER.valueToTree(
        idCollection.stream().map(UUID::toString).collect(Collectors.toList()));
  }
}

File: src/main/kotlin/net/flyingfishflash/loremlist/domain/exceptions/EntityNotFoundException.kt
package net.flyingfishflash.loremlist.domain.exceptions
import net.flyingfishflash.loremlist.domain.LrmComponentType
import net.flyingfishflash.loremlist.toJsonElement
import org.springframework.http.HttpStatus
import java.util.UUID

abstract class EntityNotFoundException(idCollection: Set<UUID>, message: String? = null, lrmComponentType: LrmComponentType? = null) :
  DomainException(
    httpStatus = HTTP_STATUS,
    title = "${lrmComponentType?.name ?: "Entity"}NotFoundException",
    message = message ?: defaultMessage(idCollection, lrmComponentType),
    supplemental = mapOf("notFound" to idCollection.map { it.toString() }.toJsonElement()),
  ) {
  companion object {
    val HTTP_STATUS = HttpStatus.NOT_FOUND
    fun defaultMessage(idCollection: Set<UUID>, lrmComponentType: LrmComponentType?): String {
      val entityName = lrmComponentType?.name ?: "Entity"
      val message = if (idCollection.size > 1) {
        "${entityName}s (${idCollection.size}) could not be found."
      } else {
        "$entityName could not be found."
      }
      return message
    }
  }
}
